### PR TITLE
Add SSL / HTTPS server test

### PR DESCRIPTION
**Pull request by FedericoCeratto**
Run a trivial HTTPS server and test its security using https://testssl.sh/
Related to #13884

### CODE REVIEW

**Comment by dom96** [comment]
Nice, looks like it timed out on ubuntu. Re-running.

**Comment by FedericoCeratto** [comment]
[DO NOT MERGE YET]

File: tests/untestable/tssl_server.nim [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  import
+    net,
+    openssl
+
+  from ssl_config import CiphersIntermediate
+
+  when defined(multithread):
+    import threadpool
+
+  when not defined(multithread):
```
Comment (konsumlamm) [inline tests/untestable/tssl_server.nim:45]:
```suggestion
  else:
```

File: .github/workflows/ci_packages.yml [hunk 2]
```diff
@@ -61,3 +61,19 @@ jobs:
       - name: 'koch, Run CI'
         shell: bash
         run: . ci/funs.sh && nimInternalBuildKochAndRunCI
+
+      - name: 'Run SSL server tests'
```
Comment (timotheecour) [inline .github/workflows/ci_packages.yml:65]: c_packages should be reserved for nimble packages though; why not handle this in azure-pipelines.yml instead which is for internal CI code?

File: tests/untestable/tssl_server.nim [hunk 3]
```diff
@@ -0,0 +1,119 @@
+discard """
+  cmd: "nim $target --threads:on -d:ssl $options $file"
```
Comment (timotheecour) [inline tests/untestable/tssl_server.nim:2]:
```suggestion
  matrix: "--threads:on -d:ssl"
```
simpler = better

File: tests/untestable/tssl_server.nim [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+"""
+
+#            Nim - Basic SSL integration tests
+#        (c) Copyright 2020 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+## Runs a simple HTTPS server
+## Used by .github/workflows/ci_ssl.yml
```
Comment (timotheecour) [inline tests/untestable/tssl_server.nim:12]:
```suggestion
```
doesn't exist anymore (and users could grep for the name to find where it's tested)

File: tests/untestable/tssl_server.nim [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+const
+  reply = "HTTP/1.0 200 OK\r\nServer: Nimtest\r\nContent-type: text/html\r\nContent-Length: 0\r\nStrict-Transport-Security: max-age=63072000; includeSubDomains\r\n\r\n"
+  certFile = "tests/stdlib/thttpclient_ssl_cert.pem"
+  keyFile = "tests/stdlib/thttpclient_ssl_key.pem"
+
+proc log(msg: string) =
+  echo "    [" & $epochTime() & "] " & msg
+
+
+when not defined(windows) and defined(sync):
```
Comment (timotheecour) [inline tests/untestable/tssl_server.nim:33]:
```suggestion
when not defined(windows) and defined(nimTestsSync):
```
(or nimTestsTssl_serverSync)
defined symbols should be prefixed even in tests

File: tests/untestable/tssl_server.nim [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import times
+
+# bogus self-signed certificate
+const
+  reply = "HTTP/1.0 200 OK\r\nServer: Nimtest\r\nContent-type: text/html\r\nContent-Length: 0\r\nStrict-Transport-Security: max-age=63072000; includeSubDomains\r\n\r\n"
+  certFile = "tests/stdlib/thttpclient_ssl_cert.pem"
+  keyFile = "tests/stdlib/thttpclient_ssl_key.pem"
+
+proc log(msg: string) =
+  echo "    [" & $epochTime() & "] " & msg
```
Comment (timotheecour) [inline tests/untestable/tssl_server.nim:30]: use strutils.%

File: tests/untestable/tssl_server.nim [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+
+when not defined(windows) and defined(sync):
+
+  from net import Port, newContext
```
Comment (timotheecour) [inline tests/untestable/tssl_server.nim:35]: dosen't make sense to have both this and import net below

File: tests/untestable/tssl_server.nim [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  import
+    net,
+    openssl
+
+  from ssl_config import CiphersIntermediate
+
+  when defined(multithread):
```
Comment (timotheecour) [inline tests/untestable/tssl_server.nim:42]: * ditto, use proper package scoping (also, that defined symbol should be referenced in the test description on top, probably even tested in CI via matrix)
* but is that even needed? IIUC all you need is:
```nim
const multithread = compileOption("threads")
...
when multithread:...
```

and then test this spec on top:
```nim
  matrix: "--threads:on ...; ..."
```

File: tests/untestable/tssl_server.nim [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    import threadpool
+
+  when not defined(multithread):
+    var ctx = newContext(protVersion = protSSLv23, verifyMode = CVerifyPeer,
+                 certFile = certFile, keyFile = keyFile, cipherList = CiphersIntermediate)
+
+  proc handleClient(client: Socket) =
+    when defined(multithread):
+      var ctx = newContext(protVersion = protSSLv23, verifyMode = CVerifyPeer,
```
Comment (timotheecour) [inline tests/untestable/tssl_server.nim:51]: let

ditto elsewhere with ref types

File: tests/untestable/tssl_server.nim [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+      close(client)
+      SSL_free(ssl)
+
+    log "closing"
+    close(socket)
+    log "server: exited"
+
+
+  when isMainModule:
+    asyncCheck runServer(8333.Port)
```
Comment (timotheecour) [inline tests/untestable/tssl_server.nim:118]: not DRY (8333 is repeated), but also, hardcoding ports is bad, see
https://github.com/timotheecour/Nim/issues/534#issuecomment-821904874

**Pull request closed without merging**